pallets/dapp-staking/src/lib.rs

```diff
@@ -2477,6 +2695,136 @@ pub mod pallet {
         }
     }
 
+    /// ActiveBonusMigration helpers
+    /// rework/reused from: https://github.com/AstarNetwork/Astar/blob/a13ef8e40ab8d26b1080236deafb8a609317f099/pallets/dapp-staking-migration/src/lib.rs
+    /// TODO: remove these helpers once BonusStatus update is done
+    impl<T: Config> Pallet<T> {
+        /// Execute update steps until the specified weight limit has been consumed.
+        ///
+        /// Depending on the number of entries updated and/or deleted, appropriate events are emitted.
+        ///
+        /// In case at least some progress is made, `Ok(_)` is returned.
+        /// If no progress is made, `Err(_)` is returned.
+        pub fn do_update(weight_limit: Weight) -> Result<Weight, Weight> {
+            // Find out if update is still in progress
+            let maybe_cursor = ActiveBonusUpdateCursor::<T>::get();
+            let mut consumed_weight = T::DbWeight::get().reads(1);
+            let mut new_cursor = None;
+            let mut entries_updated = 0u32;
+
+            let mut iter = if let Some(last_key_pair) = maybe_cursor.clone() {
+                StakerInfo::<T>::iter_from(StakerInfo::<T>::hashed_key_for(
+                    last_key_pair.0,
+                    last_key_pair.1,
+                ))
+            } else {
+                // Enable maintenance mode on the first run
+                // Likely to be already set in the runtime upgrade migration
+                ActiveProtocolState::<T>::mutate(|state| {
+                    state.maintenance = true;
```

Comment:
does this mean anyone can put protocol in maintenance mode?

Reply:
True, in a sense yes, anyone can trigger `do_update`, this is not the best. Maintenance mode is supposed to be entered with the `on_runtime_upgrade` [here](https://github.com/AstarNetwork/Astar/blob/c1fc9bb31d62285b5aa7a3aa64839d7f901cb04c/pallets/dapp-staking/src/migration.rs#L77-L81), but during chopstick testing it was not the case, so that's why I've put it also here too. I can remove it.

Reply:
I haven't noticed this before tbh - but does this mean that even after the migration finishes, this code can still be entered, and maintenance mode re-enabled?

Reply:
Btw, it's fine if anyone can trigger `do_update`, as long as it's protected properly.
* weight is limited
* once migration is finished, call has no effect

Reply:
You can test migration with chopsticks as well. Don't override wasm but instead do system::set_code and then build a block to trigger runtime upgrade then build enough blocks to run all the migrations

Reply:
I was testing it by overriding wasm, now I've followed your advice by manually building blocks, much better, thanks!